### DISCUSSION TITLE
Running Flux in Cluster A and deploying artefacts into Cluster B

### DISCUSSION THREAD

**Discussion by afritzler** [fluxcd/flux2#1867] [Q&A] [answered]
Hi Flux Community,

Following situation:

<img width="425" alt="Screenshot 2021-09-27 at 09 39 36" src="https://user-images.githubusercontent.com/1922939/134865021-73397dab-fc74-4eff-9798-692c27168757.png">

I am running a `kube-api-server` and a `kube-controller-manager` on top of a runtime Kubernetes cluster. This nodeless "cluster" is serving as a control plane for controllers which only work on their own CRDs (so no pods etc. should be scheduled there).

What would be the right way to bootstrap flux in such an environment as I can only run the `helm-controller` etc only on the underlying runtime cluster. As I need to point all the necessary flux controllers to use the nodeless cluster on top, I could of course do that by deploying them manually in the runtime cluster an set the corresponding `KUBECONFIG` `env` vars - not sure if there is a better way to achieve that.

Was curious to see if anybody else is running a similar setup.

Regards,
Andreas

**Comment by chanwit** [comment]
Yes I did a similar setup by having only a K8s controlplane and running 2 Flux controllers, Source and Kustomize outside K8s as standalone programs. What I needed to do is passing KubeConfig for them to reconcile with CRDs from that controlplane.

My app workloads in that setup were in a remote K8s cluster somewhere else (Cluster B), and it worked pretty well.

Just have KubeConfig of Cluster B as a secret in Cluster A and it would work.

I haven't tried Helm controller yet in this kind of setup, but AFAIC your setup needs at least 2 controllers, Source + X.

**Comment by stefanprodan** [comment, marked as answer]
Targeting remote clusters can be done by creating a secret with the target kubeconfig and map it to `Kustomization.spec.kubeConfig.secretRef` and `HelmRelease.spec.kubeConfig.secretRef `. Docs here:
- https://fluxcd.io/docs/components/kustomize/kustomization/#remote-clusters--cluster-api
- https://fluxcd.io/docs/components/helm/helmreleases/#remote-clusters--cluster-api

**Comment by afritzler** [comment]
Thanks @stefanprodan the `kubeConfig` ref was exactly what I was looking for.
